src/jwt/index.ts

```diff
@@ -74,22 +79,80 @@ export const isSignatureValid = (token: string, jwk: JWK): Promise<boolean> =>
+ *
+ * const { payload, protectedHeader } = await verify(jwt, jwks, {
+ *   typ: 'entity-statement+jwt',
+ *   requiredClaims: ['iss', 'sub', 'metadata'],
+ * })
+ * ```
+ *
  * @param jwt JSON Web Token value (encoded as JWS).
- * @param key Public key to verify the JWT in JWK format.
+ * @param key Public key to verify the JWT in JWK format or JWKS (array of JWK).
  * @param options JWT Decryption and JWT Claims Set validation options.
  */
 export const verify = async (
   token: string,
-  jwk: JWK,
+  jwk: JWK | JWK[],
   options: JWTClaimVerificationOptions = {}
 ): Promise<JWTDecodeResult> => {
   const { protectedHeader, payload } = await decode(token);
-  const signatureIsValid = await isSignatureValid(token, jwk);
 
+  if (Array.isArray(jwk)) {
+    jwk = getJwkFromHeader(protectedHeader, jwk);
+  }
+
+  let signatureIsValid = await isSignatureValid(token, jwk);
   if (signatureIsValid) {
     const verifiedPayload = jwtPayload(protectedHeader, payload, options);
     return { payload: verifiedPayload, protectedHeader };
   } else {
     throw new JWSSignatureVerificationFailed();
   }
 };
+
+/**
+ * Given the header of a JWT return the corresponding key from a JWKS.
+ *
+ * It uses the "alg" (JWS Algorithm) Header Parameter to determine the right JWK "kty" (Key Type),
+ * then proceeds to match the JWK "kid" (Key ID) with one found in the JWS Header Parameters (if
+ * there is one).
+ *
+ * Only a single public key must match the selection process. As shown in the example below when
+ * multiple keys get matched it is possible to opt-in to iterate over the matched keys and attempt
+ * verification in an iterative manner.
+ *
+ * @param header JWT header
+ * @param jwks Json Web Key Set (JWKS)
+ *
+ */
+export const getJwkFromHeader = (
+  header: JWSHeaderParameters,
+  jwks: JWK[]
+): JWK => {
+  let alg = header.alg;
+  let kid = header.kid;
+
+  if (alg !== undefined && kid !== undefined) {
+    let kty = getKtyFromAlg(alg);
+    let filteredJwks = jwks.filter((el) => {
+      if (el.kty === kty && el.kid === kid) {
+        return true;
+      }
+      return false;
+    });
+
+    if (filteredJwks.length === 1 && filteredJwks[0] !== undefined) {
+      return filteredJwks[0];
+    }
```

Comment:
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** If `filteredJwks.length > 1` the function throws a `JWKNotFound` exception. I see we want to have **one and only one‏** key with a given pair (kid, kty). Should we use `find` instead of `filter`, so that we only have to check for that value to no be undefined?

Reply:
great suggestion, thanks!